src/validation.cpp

```diff
@@ -4718,9 +4718,22 @@ bool Chainstate::LoadChainTip()
         return false;
     }
     m_chain.SetTip(*pindex);
+    tip = m_chain.Tip();
+
+    // Make sure our chain tip before shutting down scores better than any other candidate
+    // to maintain a consistent best tip over reboots
+    auto target = tip;
+    while (target) {
+        target->nSequenceId = 0;
+        target = target->pprev;
+    }
```

Comment:
q: cannot just set the tip's sequence id to 0?

We only compare the tip inside `TryAddBlockIndexCandidate` and not the entire chain (if not, then a test might be missing because setting only the tip to 0 still passes all tests).

Reply:
I don't think that this is an edge case we'd necessarily need to support though - it was just the only case I could come up where it would matter.

Reply:
I still think it would be sufficient to only set it for the tip - I don't think that the edge case of `invalidateblock` justifies doing another iteration over all blocks during each init (it takes 0.1 - 0.2s on mainnet for me on a reasonably fast laptop, which is not that much, but I would imagine it might be quite a bit longer with slower hardware / longer chains such as testnet3 - unless there is another reason to set it for all blocks that I am unaware of.

Reply:
IIRC, the rationale here was that `nSequenceId` is supposed to be monotonically increasing, so having a chain of blocks where the previous blocks all have nSequenceId = `x-1` and the tip has `x` was weird. Since this is something that should only happen on restarts, the extra fraction of a second was not that relevant.

I'm happy to change it so only the tip has 0 if you think the time savings are more relevant than the inconsistency with the design. Comments will also need to be updated.

Reply:
> so having a chain of blocks where the previous blocks all have nSequenceId = x-1 and the tip has x

Sorry, I wrote that wrong.  `:s/x-1/x+1/g`

In the case were we only change the tip, the rest of blocks will have `nSequenceId=1`, so we will end up with a chain like:

`1 -> 1 -> 1 -> 0 (tip)` that will evolve to `1 -> 1 -> 1 -> 0 -> 2 (tip)`

All other chains will have `nSequenceId=1` from start to end.

In the case proposed here, we'll have 

`0 -> 0 -> 0 -> 0 (tip)` evolving to `0 -> 0 -> 0 -> 0 -> 2 (tip)`

Reply:
oh right, I got confused in my earlier post, you are right. Still, maybe others could chime in if they think that this monotonicity is worth the added work or not.